File: 34.py
class Solution:
    # @param {integer[]} nums
    # @param {integer} target
    # @return {integer[]}
    def searchRange(self, nums, target):
        n = len(nums)
        if n == 0:
            return [-1,-1]
        if n == 1:
            if nums[0] == target:
                return [0,0]
            else:
                return [-1,-1]
        mid = n/2
        mid_val = nums[mid]
        if target < mid_val:
            the_range = self.searchRange(nums[0:mid], target)
            return the_range
        elif target > mid_val:
            the_range = self.searchRange(nums[mid+1:], target)
            if the_range[0] == -1:
                return the_range
            else:
                return [the_range[0]+mid+1, the_range[1]+mid+1]
        else:
            start_idx = mid - 1
            while start_idx >= 0:
                if nums[start_idx] == target:
                    start_idx -= 1
                else:
                    break
            start_idx += 1
            
            end_idx = mid + 1
            while end_idx <= n-1 :
                if nums[end_idx] == target:
                    end_idx += 1
                else:
                    break
            end_idx -= 1   
            
            return [start_idx,end_idx]
        
        

File: 34.cpp
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        int n = nums.size();
        int low = 0;
        int high = n - 1;
        int mid = 0;
        int index = -1;
        while (low <= high){
            mid = low + (high - low)/2;
            if(nums[mid] < target)
                low = mid+1;
            else if (nums[mid] == target)
            {
                index = mid;
                break;
            }
            else
                high = mid - 1;
        }
        
        if (index != -1){
            int startIndex = mid;
            while(startIndex>0 && nums[startIndex-1] == target)
                startIndex--;
            int endIndex = mid;
            while(endIndex<n-1 && nums[endIndex+1] == target)
                endIndex++;
            return vector<int>{startIndex,endIndex};
        }
        else
            return vector<int>{-1,-1};
    }
};